frontend/src/components/data-table/column-header.tsx

```diff
@@ -522,23 +546,214 @@ const TextFilter = <TData, TValue>({
+  };
+
+  return (
+    <Popover open={true}>
+      <PopoverTrigger />
+      <PopoverContent>
+        <PopoverClose className="absolute top-2 right-2">
+          <Button
+            variant="link"
+            size="sm"
+            onClick={() => setSetFilterOpen(false)}
+          >
+            X
+          </Button>
+        </PopoverClose>
+        <div className="flex flex-col gap-1.5 pt-1">
+          <span className="text-sm font-semibold mx-auto mb-2">Set Filter</span>
+          <Input
+            value={query}
+            onChange={(e) => setQuery(e.target.value)}
+            placeholder="Search"
+            className="my-0 py-1"
+            autoFocus={true}
+          />
+          <Table className="w-full border-collapse text-sm overflow-auto max-h-60 ">
+            <TableHeader>
+              <TableRow className="hover:bg-transparent">
+                <TableHead>{}</TableHead>
+                {keys.map((key) => (
+                  <TableHead key={key} className="text-foreground">
+                    {key}
+                  </TableHead>
+                ))}
+              </TableRow>
+            </TableHeader>
+            <TableBody>
+              {filteredData.map((row, rowIndex) => {
+                const value = row[column.id];
+                return (
+                  <TableRow key={rowIndex}>
+                    <TableCell>
+                      <Checkbox
+                        onCheckedChange={(checked) => {
+                          if (typeof checked === "string") {
+                            return;
+                          }
+                          handleCheckboxClick(checked, value);
+                        }}
+                        aria-label="Select row"
+                      />
+                    </TableCell>
+                    {keys.map((key) => (
+                      <TableCell key={`${rowIndex}-${key}`}>
+                        {row[key] === null ? "null" : String(row[key])}
+                      </TableCell>
+                    ))}
+                  </TableRow>
+                );
+              })}
+            </TableBody>
+          </Table>
+          <FilterButtons
+            onApply={handleApply}
+            onClear={() => {
+              setChosenValues([]);
+            }}
+            clearButtonDisabled={chosenValues.length === 0}
+          />
+        </div>
+      </PopoverContent>
+    </Popover>
+  );
+};
+
+const FilterButtons = ({
+  onApply,
+  onClear,
+  clearButtonDisabled,
+}: {
+  onApply: () => void;
+  onClear: () => void;
+  clearButtonDisabled?: boolean;
+}) => {
+  return (
+    <div className="flex gap-2 px-2 justify-between">
+      <Button variant="link" size="sm" onClick={onApply}>
+        Apply
+      </Button>
+      <Button
+        variant="linkDestructive"
+        size="sm"
+        className=""
+        onClick={onClear}
+        disabled={clearButtonDisabled}
+      >
+        Clear
+      </Button>
```

Comment:
could this be refactored into another file?

also, i think you can use `column.getFilterValue` and `column.setFilterValue`. here are the APIs: https://tanstack.com/table/v8/docs/guide/column-filtering#column-filter-apis